### PR TITLE
Sifive uart receive

### PR DESCRIPTION
**Pull request by mateibarbu19**
### Pull Request Overview

This pull request adds receive functionality to the UART HIL implementation for the SiFive chips.

This is done similarly to the `rp2040` chip. 

### Testing Strategy

By adding a `ProcessConsole` to both of the HiFive boards. (I tested the `hifive1` board in QEMU). On the Inventor I flashed the following [example](https://github.com/tock/libtock-c/tree/master/examples) applications from `libtock-c`:

- [x] `c_hello`
- [x] `console`
- [x] `console_recv_long`
- [x] `console_recv_short`

### Help Wanted

This pull request still needs to be tested, I think, with some applications on the `hifive1` board.

@gemarcano could you please do this?

### Documentation Updated

- [x] Updated the relevant files in `boards/README.md`

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

**Review by bradjc — APPROVED** [review]

File: boards/hifive1/src/main.rs [hunk 1]
```diff
@@ -237,6 +237,17 @@ pub unsafe fn main() {
         components::process_printer::ProcessPrinterTextComponent::new().finalize(());
     PROCESS_PRINTER = Some(process_printer);
 
+    let _process_console = components::process_console::ProcessConsoleComponent::new(
```
Comment (alistair23) [inline boards/hifive1/src/main.rs:240]: Can you remove the `_` before `_process_console`?

File: boards/hifive_inventor/src/main.rs [hunk 2]
```diff
@@ -216,6 +216,17 @@ pub unsafe fn main() {
         components::process_printer::ProcessPrinterTextComponent::new().finalize(());
     PROCESS_PRINTER = Some(process_printer);
 
+    let _process_console = components::process_console::ProcessConsoleComponent::new(
```
Comment (alistair23) [inline boards/hifive_inventor/src/main.rs:219]: Same here

**Review by alistair23 — DISMISSED** [review]

**Review by twilfredo — APPROVED** [review]

**Review by bradjc — APPROVED** [review]

**Review by alexandruradovici — APPROVED** [review]

**Review by alistair23 — APPROVED** [review]

**Review by twilfredo — APPROVED** [review]

**Comment by ppannuto** [comment]
bors d=gemarcano 

@gemarcano can you test this on hw when you have a moment, and 'r+' if all is well?

**Comment by gemarcano** [comment]
It does look like UART receive is working. I am seeing the main console crash if I do a `list` while an app is running, however, but it might be unrelated to this PR.

Steps to reproduce on hardware:
1. build and flash c_hello
2. on UART console (via tockloader listen), type `list` and press enter

Expected:
Some sort of list of applications?

Observed:
Panic

```
$ tockloader listen
[INFO   ] No device name specified. Using default name "tock".
[INFO   ] No serial port with device name "tock" found.
[INFO   ] Found 2 serial ports.
Multiple serial port options found. Which would you like to use?
[0]     /dev/ttyACM1 - HiFive - CDC
[1]     /dev/ttyACM0 - HiFive - CDC

Which option? [0] 1
[INFO   ] Using "/dev/ttyACM0 - HiFive - CDC".
[INFO   ] Listening for serial output.
␛[1F␛[0J␛[?25h␛[0;39;49mBench Clock Reset Complete

ATE0-->OK
AT+BLEINIT=0-->OK
AT+CWMODE=0-->OK

HiFive1 initialization complete. Entering main loop.
Hello World!
tock$ list

panicked at 'fatal exception', chips/e310x/src/chip.rs:189:13
        Kernel version release-2.1-94-g25b545f7c

---| No debug queue found. You can set it with the DebugQueue component.

---| RISC-V Machine State |---
Last cause (mcause): Store/AMO access fault (interrupt=0, exception code=0x00000007)
Last value (mtval):  0x7FFFFDA4

System register dump:
 mepc:    0x20014036    mstatus:     0x00001880
 mcycle:  0x07971DE2    minstret:    0x00B514AA
 mtvec:   0x20010100
 mstatus: 0x00001880
  uie:    false  upie:   false
  sie:    false  spie:   false
  mie:    false  mpie:   true
  spp:    false
 mie:   0x00000888   mip:   0x00000000
  usoft:  false               false 
  ssoft:  false               false 
  msoft:  true                false 
  utimer: false               false 
  stimer: false               false 
  mtimer: true                false 
  uext:   false               false 
  sext:   false               false 
  mext:   true                false 

---| App Status |---
𝐀𝐩𝐩: c_hello   -   [Yielded]
 Events Queued: 0   Syscall Count: 8   Dropped Upcall Count: 0
 Restart Count: 0
 Last Syscall: Yield { which: 1, address: 0x0 }
 Completion Code: None


 ╔═══════════╤══════════════════════════════════════════╗
 ║  Address  │ Region Name    Used | Allocated (bytes)  ║
 ╚0x80003E94═╪══════════════════════════════════════════╝
             │ Grant Ptrs       24
             │ Upcalls         360
             │ Process         824
  0x800039DC ┼───────────────────────────────────────────
             │ ▼ Grant          76
  0x80003990 ┼───────────────────────────────────────────
             │ Unused
  0x800031DC ┼───────────────────────────────────────────
             │ ▲ Heap            0 |   1972               S
  0x800031DC ┼─────────────────────────────────────────── R
             │ Data            476 |    476               A
  0x80003000 ┼─────────────────────────────────────────── M
             │ ▼ Stack         112 |   2048          
  0x80002F90 ┼───────────────────────────────────────────
             │ Unused
  0x80002800 ┴───────────────────────────────────────────
             .....
  0x20040484 ┬─────────────────────────────────────────── F
             │ App Flash      1060                        L
  0x20040060 ┼─────────────────────────────────────────── A
             │ Protected        96                        S
  0x20040000 ┴─────────────────────────────────────────── H

 R0 : 0x00000000    R16: 0x00000000
 R1 : 0x20040274    R17: 0x00000000
 R2 : 0x80002FE0    R18: 0x80002800
 R3 : 0x00000000    R19: 0x00000000
 R4 : 0x00000000    R20: 0x00000000
 R5 : 0x80003000    R21: 0x00000000
 R6 : 0x80002FA0    R22: 0x00000000
 R7 : 0x00000000    R23: 0x00000000
 R8 : 0x20040060    R24: 0x00000000
 R9 : 0x80002800    R25: 0x00000000
 R10: 0x00000001    R26: 0x00000000
 R11: 0x00000000    R27: 0x00000000
 R12: 0x80003000    R28: 0x00000000
 R13: 0x00000000    R29: 0x00000000
 R14: 0x00000000    R30: 0x00000000
 R15: 0x00000000    R31: 0x00000000
 PC : 0x20040280

 mcause: 0x00000008 (Environment call from U-mode)
 mtval:  0x00000000


 Total number of grant regions defined: 3
  Grant  0 : --          Grant  1 0x1: 0x80003990  Grant  2 : --        
 PMP regions:
  [0]: addr=0x20040000, size=0x00000484, cfg=0xD (r-x)
  [1]: addr=0x80002800, size=0x000009DC, cfg=0xB (rw-)
  <unset>
  <unset>

To debug, run `make lst` in the app's folder
and open the arch.0x20040060.0x80002800.lst file.
```

I'll try to do some debugging to see what's at fault. If it's not the UART code, I'll approve this PR as the receive functionality does appear to be working.

**Comment by gemarcano** [comment]
The faults I'm finding are most likely the fault of something else, as other commands do work as intended, and the examples do appear to be working as intended as well.

**Pull request merged**